components/core/src/clp/streaming_compression/lzma/Compressor.cpp

```diff
@@ -0,0 +1,215 @@
+    char const* msg{nullptr};
+    switch (rc) {
+        case LZMA_MEM_ERROR:
+            msg = "Memory allocation failed";
+            break;
+
+        case LZMA_OPTIONS_ERROR:
+            msg = "Specified preset is not supported";
+            break;
+
+        case LZMA_UNSUPPORTED_CHECK:
+            msg = "Specified integrity check is not supported";
+            break;
+
+        default:
+            // This is most likely LZMA_PROG_ERROR indicating a bug in liblzma
+            msg = "Unknown error";
+            break;
+    }
+
+    SPDLOG_ERROR("Error initializing the encoder: {} (error code {})", msg, static_cast<int>(rc));
+    throw Compressor::OperationFailed(clp::ErrorCode_BadParam, __FILENAME__, __LINE__);
+}
+}  // namespace
+
+namespace clp::streaming_compression::lzma {
+auto Compressor::open(FileWriter& file_writer, int compression_level) -> void {
+    if (nullptr != m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotReady, __FILENAME__, __LINE__);
+    }
+
+    if (compression_level < cMinCompressionLevel || compression_level > cMaxCompressionLevel) {
+        throw OperationFailed(ErrorCode_Unsupported, __FILENAME__, __LINE__);
+    }
+
+    m_compression_stream = LZMA_STREAM_INIT;
+    init_lzma_encoder(&m_compression_stream, compression_level, m_dict_size);
+
+    // No input upon initialization
+    m_compression_stream.next_in = nullptr;
+    m_compression_stream.avail_in = 0;
+
+    // Attach output buffer to LZMA stream
+    m_compression_stream.next_out = m_compressed_stream_block_buffer.data();
+    m_compression_stream.avail_out = m_compressed_stream_block_buffer.size();
+
+    m_compressed_stream_file_writer = &file_writer;
+
+    m_uncompressed_stream_pos = 0;
+}
+
+auto Compressor::close() -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    run_lzma(LZMA_FINISH);
+    lzma_end(&m_compression_stream);
+
+    // Detach output buffer from LZMA stream
+    m_compression_stream.next_out = nullptr;
+    m_compression_stream.avail_out = 0;
+
+    m_compressed_stream_file_writer = nullptr;
+}
+
+auto Compressor::write(char const* data, size_t data_length) -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    if (0 == data_length) {
+        // Nothing needs to be done because we do not need to compress anything
+        return;
+    }
+
+    if (nullptr == data) {
+        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+
+    // Attach input data to LZMA stream
+    m_compression_stream.next_in = clp::size_checked_pointer_cast<uint8_t const>(data);
+    m_compression_stream.avail_in = data_length;
+
+    run_lzma(LZMA_RUN);
+
+    m_uncompressed_stream_pos += data_length;
+}
+
+auto Compressor::flush() -> void {
+    if (m_compression_stream_is_flushed) {
```

Comment:
If we separate the functions run_lzma into two functions, maybe this line can also be moved into one of them.